src/software/ai/navigator/path_planner/enlsvg_path_planner.cpp

```diff
@@ -0,0 +1,199 @@
+                          new_end.value().x, new_end.value().y);
+    std::optional<Path> path = convertEnlsvgPathToPath(enlsvgPath);
+    if (path == std::nullopt)
+    {
+        return std::nullopt;
+    }
+
+    std::vector<Point> path_points = path.value().getKnots();
+
+    // If start was initially blocked, add the start point
+    if (new_start.value() != s)
+    {
+        path_points.insert(path_points.begin(), start);
+    }
+
+    // If the end point wasn't blocked, then replace the end with the actual end because
+    // some details get lose due to the grid resolution
+    if (new_end.value() == e)
+    {
+        path_points.pop_back();
+        path_points.emplace_back(end);
+    }
+
+    // Make sure start point corresponds exactly with the given start
+    path_points.erase(path_points.begin());
+    path_points.insert(path_points.begin(), start);
+
+    // Due to processing, it is possible that the first two points may be very close
+    // together, this will fix that
+    if (path_points.size() > 2 && (path_points[0] - path_points[1]).length() < resolution)
+    {
+        path_points.erase(path_points.begin() + 1);
+    }
+
+    return Path(path_points);
+}
+
+EnlsvgPathPlanner::EnlsvgPoint EnlsvgPathPlanner::convertPointToEnlsvgPoint(
+    const Point &p) const
+{
+    return EnlsvgPoint(
+        static_cast<int>(round((p.x() - origin.x()) / (double)resolution)),
+        static_cast<int>(round((p.y() - origin.y()) / (double)resolution)));
+}
+
+Point EnlsvgPathPlanner::convertEnlsvgPointToPoint(const EnlsvgPoint &c) const
+{
+    return Point(static_cast<double>(c.x * resolution + origin.x()),
+                 static_cast<double>(c.y * resolution + origin.y()));
+}
+
+std::optional<Path> EnlsvgPathPlanner::convertEnlsvgPathToPath(const EnlsvgPath &p) const
+{
+    if (!p.size())
+    {
+        return std::nullopt;
+    }
+
+    std::vector<Point> path;
+    for (auto gv : p)
+    {
+        path.emplace_back(convertEnlsvgPointToPoint(gv));
+    }
+
+    return std::optional(Path(path));
+}
+
+std::optional<EnlsvgPathPlanner::EnlsvgPoint>
+EnlsvgPathPlanner::findClosestUnblockedEnlsvgPoint(const EnlsvgPoint &ep) const
+{
+    // Uses DFS to find the closest unblocked cell by looking at nearby cells
+    std::queue<EnlsvgPoint> q;
+    std::vector<EnlsvgPoint> visited;
+    q.push(ep);
+    visited.emplace_back(ep);
+    while (!q.empty())
+    {
+        EnlsvgPoint test_coord = q.front();
+        q.pop();
+        if (!isBlocked(test_coord))
+        {
+            return std::optional<EnlsvgPoint>(test_coord);
+        }
+        for (int i = 1; i > -2; i -= 2)
+        {
+            EnlsvgPoint next_coord = EnlsvgPoint(test_coord.x + i, test_coord.y);
+            for (int j = -1; j < 2; j += 2)
+            {
+                if (isCoordNavigable(next_coord) &&
+                    (std::find(visited.begin(), visited.end(), next_coord) ==
+                     visited.end()))
```

Comment:
it looks like the order you visit doesn't really matter

so you can use std::unordered_set instead of std::vector

should be able to do `set.count(next_coord) == 1` here

Reply:
changed